csrc/val_graph_visitor.cpp

```diff
@@ -0,0 +1,126 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <val_graph_visitor.h>
+
+#include <id_model/to_string.h>
+
+namespace nvfuser {
+
+void ValGraphVisitor::traverse() {
+  const ValGroups terminating_inputs = graph().getTerminatingInputs();
+  std::deque<ValGroup> to_visit_vals(
+      terminating_inputs.begin(), terminating_inputs.end());
+  ValGroups visited_vals;
+
+  std::deque<ExprGroup> to_visit_exprs;
+  ExprGroups visited_exprs;
+
+  auto is_expr_ready = [&](const ExprGroup& expr_group) -> bool {
+    const auto inp_groups = graph().inputGroups(expr_group);
+    return std::all_of(
+        inp_groups.begin(), inp_groups.end(), [&](ValGroup val_group) {
+          return visited_vals.has(val_group) || val_group->empty();
+        });
+  };
+
+  // If any input of the def expr is mapped with the val
+  // group itself, i.e., a trivial expr, allow visiting the
+  // val group first. The trivial expr group will be visited
+  // after the val group.
+  //
+  // Example:
+  //
+  // [i0, 1]
+  // merge
+  // [i0*1]
+  // map i0 and i0*1
+  // ValGroups: {{i0, i0*1}, {1}}
+  //
+  // Then, {i0, i0*1} and {1} would be visited first, then the merge
+  // expr group would be visited. {i0, i0*1} is also an output group
+  // of the merge but since it's already in the visited set, it would
+  // not be visited again.
+  //
+  // See also IdModelTest.ValGraphStmtSort3 for a concrete example.
+  auto is_val_ready = [&](const ValGroup& val_group) -> bool {
+    const ExprGroups& unique_defs = graph().getDefinitions(val_group);
+    return std::all_of(
+        unique_defs.begin(), unique_defs.end(), [&](ExprGroup expr_group) {
+          return expr_group->empty() || visited_exprs.has(expr_group) ||
+              terminating_inputs.has(val_group) ||
+              graph().isTrivialExprGroup(expr_group);
+        });
+  };
+
+  while (!to_visit_vals.empty() || !to_visit_exprs.empty()) {
+    // Process expressions first as all definitions of vals have to be
+    // processed before we can process that val.
+
+    // Detect if nothing has been processed which would put us in an infinite
+    // loop
+    bool something_was_processed = false;
+    std::deque<ExprGroup> still_to_visit_exprs;
```

Comment:
I think we can get rid of `still_to_visit_exprs`. Let me know if the following makes sense and it's certainly fine to do it in a separate PR. 

All we need is to guarantee `to_visit_*` contains only the ready ones. An unready expr/val can be attempted multiple times to be added to `to_visit_*`, but only when its last-to-visit operand is handled it's actually added to `to_visit_*`. This probably saves some time because an expr/val don't need to be bounced between `to_visit_*` and `still_to_visit_*` potentially many times.

Reply:
I think you're right. Let me clean it up.

Reply:
(As I said, it's certainly fine to do it in a separate PR. I mainly wanted to run this idea through you to make sure I understand your code correctly.)

Reply:
Updated: b92ec79cfa013dbc8512afc1e0157f6c36f723cd